server/controllers/frameController.js

```diff
@@ -0,0 +1,35 @@
+import { Frame } from "../models/index.js";
+import { NotFound } from "../utils/httpError.js";
+
+export const getAllFrames = async (req, res) => {
+  const frames = await Frame.find();
+  res.status(200).json(frames);
+  if (!frames) {
+    throw new NotFound("No frames found");
+  }
+};
+
+export const getFrameById = async (req, res) => {
+  const { id } = req.params;
+  const frame = await Frame.findById(id);
+  res.status(200).json(frame);
+  if (!frame) {
+    throw new NotFound("Frame with id not found");
+  }
+};
+
+export const getRandomFrameForLevel = async (req, res) => {
+    const { levelId } = req.params;
+    const levelExists = await Frame.exists({ difficultyLevel: levelId });
+    if (!levelExists) {
+      throw new NotFound("Level does not exist");
+    }
```

Comment:
The second query (`Frame.find()`) already implicitly checks if any documents match the levelId in the difficultyLevel field. If no documents match, the result will be an empty array (`frames.length === 0`), which is handled by the second if block.

So, the first check with `Frame.exists()` is redundant because the second query essentially does the same thing. Or that or I have misunderstood your point here.

Reply:
I was thinking the first one is to check if the level exists in the database and the second one is to check if the existed level has any frames. 
```
export const getRandomFrameForLevel = async (req, res) => {
    const { levelId } = req.params;
    const levelExists = await Frame.exists({ difficultyLevel: levelId });
    if (!levelExists) {
      throw new NotFound("Level does not exist"); // if the level does not exist in the database
    }
    else {
        const frames = await Frame.find({ difficultyLevel: { $in: [levelId] } }); 
        if (frames.length === 0) {
            throw new NotFound("No frames found"); // if the level does not have any frames
        }
        const randomFrame = frames[Math.floor(Math.random() * frames.length)]; 
        res.status(200).json(randomFrame); 
    }
};
```

Reply:
If you want to check whether or not a level exists in the database, then it should be check by calling the level model, not by calling frame model and check the inner level fields. 

For example, if there is no level 10 in the database, but Frame has one object which includes 
```
{..., 
difficultyLevel: [1, 2, 3, 10], 
...} 
```
then the error won't be thrown at all (which is not what we want?).

Call this instead if you wanna check for the avalibility of the level:

```
const level = await Level.findById(levelId); 
if (!level) throw new NotFound("Level does not exist"); 
```